GraphQL error: No operation named ""

Was trying out the Apollo Client for the most basic of examples (from README) to get data from Github API but have been hitting a strange error.

**Intended outcome:**
- Data from Github API

**Actual outcome:**
No data, error instead:
```
proxyConsole.js:56 Error: GraphQL error: No operation named ""
    at new ApolloError (http://localhost:3000/static/js/bundle.js:60113:28)
    at http://localhost:3000/static/js/bundle.js:62408:41
    at http://localhost:3000/static/js/bundle.js:62845:25
    at Array.forEach (native)
    at http://localhost:3000/static/js/bundle.js:62842:27
    at Array.forEach (native)
    at QueryManager.broadcastQueries (http://localhost:3000/static/js/bundle.js:62839:42)
    at Array.<anonymous> (http://localhost:3000/static/js/bundle.js:62205:23)
    at dispatch (http://localhost:3000/static/js/bundle.js:61742:19)
    at http://localhost:3000/static/js/bundle.js:62056:30
    at Object.dispatch (http://localhost:3000/static/js/bundle.js:60224:16)
    at http://localhost:3000/static/js/bundle.js:62783:29
    at <anonymous>
```

**How to reproduce the issue:**
Code; I used:
```
// networkInterface has Github Personal Bearer Token for authentication/authorization
const apollo_client = new ApolloClient({
    networkInterface,
})

apollo_client.query({
    
    query: gql`{
        viewer {
            avatarUrl
        }
    }`,

}).then(data => console.log(data)).catch(error => console.error(error))

```

Raw data sent over the wire:
```
{"query":"{\n  viewer {\n    avatarUrl\n    __typename\n  }\n}\n","operationName":""}
```

Have tried to put operationName in but same result:
```
apollo_client.query({
    
    query: gql`{
        viewer {
            avatarUrl
        }
    }`,
    operationName: 'Query1'

}).then(data => console.log(data)).catch(error => console.error(error))

```

I found out I need to manipulate the operationName directly to set it to null to get things working:
```
networkInterface.use([{

    applyMiddleware(req, next) {

        if (!req.options.headers) {
            req.options.headers = {}
        }
        const token = "abcdef"
        req.options.headers.authorization = token ? `Bearer ${token}` : null

        req.request.variables = {}
        req.request.operationName = null

        next()
    }

}])

```

Should operationName be set to null by default instead of "" if no value was passed in? And how to set operationName correctly (is it really needed to be manipulated in the middleware)?

Hope some guidance can be provided.  Thanks!

Thanks @stubailo for the tips, I did the below change (total TS newbie) and tried to compile:
```
export function getOperationName(doc: DocumentNode): string | null {
  let res: string | null = null;
  doc.definitions.forEach((definition) => {
..
```

but encountered the following scary output -- not sure how to fix and where the rabbit hole will end 🐰 :
```
> apollo-client@1.4.0 compile /Users/leow/Desktop/PROJECTS/APOLLO/apollo-client
> tsc

src/core/ObservableQuery.ts(415,40): error TS2345: Argument of type '{ type: "APOLLO_UPDATE_QUERY_RESULT"; newResult: any; variables: { [key: string]: any; } | undefi...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_UPDATE_QUERY_RESULT"; newResult: any; variables: { [key: string]: any; } | undefi...' is not assignable to type 'UpdateQueryResultAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/core/QueryManager.ts(285,25): error TS2345: Argument of type '{ type: "APOLLO_MUTATION_INIT"; mutationString: string; mutation: DocumentNode; variables: Object...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_MUTATION_INIT"; mutationString: string; mutation: DocumentNode; variables: Object...' is not assignable to type 'MutationInitAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/core/QueryManager.ts(316,31): error TS2345: Argument of type '{ type: "APOLLO_MUTATION_RESULT"; result: ExecutionResult; mutationId: string; document: Document...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_MUTATION_RESULT"; result: ExecutionResult; mutationId: string; document: Document...' is not assignable to type 'MutationResultAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/core/QueryManager.ts(419,25): error TS2345: Argument of type '{ type: "APOLLO_QUERY_INIT"; queryString: string; document: DocumentNode; operationName: string |...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_QUERY_INIT"; queryString: string; document: DocumentNode; operationName: string |...' is not assignable to type 'QueryInitAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/core/QueryManager.ts(441,27): error TS2345: Argument of type '{ type: "APOLLO_QUERY_RESULT_CLIENT"; result: { data: any; }; variables: { [key: string]: any; };...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_QUERY_RESULT_CLIENT"; result: { data: any; }; variables: { [key: string]: any; };...' is not assignable to type 'QueryResultClientAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/core/QueryManager.ts(863,11): error TS2322: Type '{ query: DocumentNode; variables: ({ [key: string]: any; } & undefined) | ({ [key: string]: any; ...' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
src/core/QueryManager.ts(887,33): error TS2345: Argument of type '{ type: "APOLLO_SUBSCRIPTION_RESULT"; document: DocumentNode; operationName: string | null; resul...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_SUBSCRIPTION_RESULT"; document: DocumentNode; operationName: string | null; resul...' is not assignable to type 'SubscriptionResultAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/core/QueryManager.ts(1063,11): error TS2322: Type '{ query: DocumentNode; variables: { [key: string]: any; } | undefined; operationName: string | nu...' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
src/core/QueryManager.ts(1078,31): error TS2345: Argument of type '{ type: "APOLLO_QUERY_RESULT"; document: DocumentNode; operationName: string | null; result: Exec...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_QUERY_RESULT"; document: DocumentNode; operationName: string | null; result: Exec...' is not assignable to type 'QueryResultAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
src/data/proxy.ts(242,25): error TS2345: Argument of type '{ type: "APOLLO_WRITE"; writes: { rootId: string; result: any; document: DocumentNode; operationN...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_WRITE"; writes: { rootId: string; result: any; document: DocumentNode; operationN...' is not assignable to type 'WriteAction'.
    Types of property 'writes' are incompatible.
      Type '{ rootId: string; result: any; document: DocumentNode; operationName: string | null; variables: O...' is not assignable to type 'DataWrite[]'.
        Type '{ rootId: string; result: any; document: DocumentNode; operationName: string | null; variables: O...' is not assignable to type 'DataWrite'.
          Types of property 'operationName' are incompatible.
            Type 'string | null' is not assignable to type 'string'.
              Type 'null' is not assignable to type 'string'.
src/data/proxy.ts(271,25): error TS2345: Argument of type '{ type: "APOLLO_WRITE"; writes: { rootId: string; result: any; document: DocumentNode; operationN...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_WRITE"; writes: { rootId: string; result: any; document: DocumentNode; operationN...' is not assignable to type 'WriteAction'.
    Types of property 'writes' are incompatible.
      Type '{ rootId: string; result: any; document: DocumentNode; operationName: string | null; variables: O...' is not assignable to type 'DataWrite[]'.
        Type '{ rootId: string; result: any; document: DocumentNode; operationName: string | null; variables: O...' is not assignable to type 'DataWrite'.
          Types of property 'operationName' are incompatible.
            Type 'string | null' is not assignable to type 'string'.
              Type 'null' is not assignable to type 'string'.
src/data/proxy.ts(410,21): error TS2345: Argument of type '{ rootId: string; result: any; document: DocumentNode; operationName: string | null; variables: O...' is not assignable to parameter of type 'DataWrite'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string'.
      Type 'null' is not assignable to type 'string'.
src/data/proxy.ts(439,21): error TS2345: Argument of type '{ rootId: string; result: any; document: DocumentNode; operationName: string | null; variables: O...' is not assignable to parameter of type 'DataWrite'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string'.
      Type 'null' is not assignable to type 'string'.
src/data/store.ts(180,91): error TS2345: Argument of type '{ mutationResult: ExecutionResult; queryName: string | null; queryVariables: Object; }' is not assignable to parameter of type '{ mutationResult: Object; queryName: Object; queryVariables: Object; }'.
  Types of property 'queryName' are incompatible.
    Type 'string | null' is not assignable to type 'Object'.
      Type 'null' is not assignable to type 'Object'.
test/deduplicator.ts(20,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/deduplicator.ts(26,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/deduplicator.ts(55,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/deduplicator.ts(97,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/deduplicator.ts(103,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/deduplicator.ts(133,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/deduplicator.ts(139,11): error TS2322: Type '{ query: DocumentNode; variables: { x: string; }; operationName: string | null; }' is not assignable to type 'Request'.
  Types of property 'operationName' are incompatible.
    Type 'string | null' is not assignable to type 'string | undefined'.
      Type 'null' is not assignable to type 'string | undefined'.
test/store.ts(177,20): error TS2345: Argument of type '{ type: "APOLLO_QUERY_INIT"; queryId: string; queryString: string; document: any; operationName: ...' is not assignable to parameter of type 'ApolloAction'.
  Type '{ type: "APOLLO_QUERY_INIT"; queryId: string; queryString: string; document: any; operationName: ...' is not assignable to type 'QueryInitAction'.
    Types of property 'operationName' are incompatible.
      Type 'string | null' is not assignable to type 'string'.
        Type 'null' is not assignable to type 'string'.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! apollo-client@1.4.0 compile: `tsc`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the apollo-client@1.4.0 compile script.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 
 ### vNEXT
 - Fix: broken edge case when setting up fragment matching with Typescript by fixing types on `IntrospectionResultData` [PR #1763](https://github.com/apollographql/apollo-client/pull/1763)
+- Fix: getOperationName now returns null when no operation name can be found in the document [PR #1769](https://github.com/apollographql/apollo-client/pull/1769)
 
 ### v1.4.0
 - Feature: Add `operationName` to Redux actions where possible [PR #1676](https://github.com/apollographql/apollo-client/pull/1676)

--- a/index.js.flow
+++ b/index.js.flow
@@ -76,14 +76,14 @@ declare module "apollo-client" {
     debugName?: string,
     query?: DocumentNode,
     variables?: Object,
-    operationName?: string,
+    operationName?: string | null,
   }
 
   declare export interface PrintedRequest {
     debugName?: string,
     query?: string,
     variables?: Object,
-    operationName?: string,
+    operationName?: string | null,
   }
 
   declare export interface SubscriptionNetworkInterface {
@@ -356,7 +356,7 @@ declare module "apollo-client" {
     previousResult: Object,
     options: {
       mutationResult: Object,
-      queryName: Object,
+      queryName: string | null,
       queryVariables: Object,
     },
   ) => Object;
@@ -585,7 +585,7 @@ declare module "apollo-client" {
     result: ExecutionResult,
     queryId: string,
     document: DocumentNode,
-    operationName: string,
+    operationName: string | null,
     requestId: number,
     fetchMoreForQueryId?: string,
     extraReducers?: ApolloReducer[],
@@ -651,7 +651,7 @@ declare module "apollo-client" {
     mutationString: string,
     mutation: DocumentNode,
     variables: Object,
-    operationName: string,
+    operationName: string | null,
     mutationId: string,
     optimisticResponse: Object | void,
     extraReducers?: ApolloReducer[],
@@ -665,7 +665,7 @@ declare module "apollo-client" {
     type: "APOLLO_MUTATION_RESULT",
     result: ExecutionResult,
     document: DocumentNode,
-    operationName: string,
+    operationName: string | null,
     variables: Object,
     mutationId: string,
     extraReducers?: ApolloReducer[],
@@ -699,7 +699,7 @@ declare module "apollo-client" {
     subscriptionId: number,
     variables: Object,
     document: DocumentNode,
-    operationName: string,
+    operationName: string | null,
     extraReducers?: ApolloReducer[],
   }
 

--- a/src/actions.ts
+++ b/src/actions.ts
@@ -24,7 +24,7 @@ export type QueryResultAction = {
   result: ExecutionResult;
   queryId: string;
   document: DocumentNode;
-  operationName: string;
+  operationName: string | null;
   requestId: number;
   fetchMoreForQueryId?: string;
   extraReducers?: ApolloReducer[];
@@ -50,7 +50,7 @@ export interface QueryInitAction {
   type: 'APOLLO_QUERY_INIT';
   queryString: string;
   document: DocumentNode;
-  operationName: string;
+  operationName: string | null;
   variables: Object;
   fetchPolicy: FetchPolicy;
   queryId: string;
@@ -69,7 +69,7 @@ export function isQueryInitAction(action: ApolloAction): action is QueryInitActi
 export interface QueryResultClientAction {
   type: 'APOLLO_QUERY_RESULT_CLIENT';
   result: ExecutionResult;
-  operationName: string;
+  operationName: string | null;
   complete: boolean;
   queryId: string;
   requestId: number;
@@ -93,7 +93,7 @@ export interface MutationInitAction {
   mutationString: string;
   mutation: DocumentNode;
   variables: Object;
-  operationName: string;
+  operationName: string | null;
   mutationId: string;
   optimisticResponse: Object | Function | undefined;
   extraReducers?: ApolloReducer[];
@@ -110,7 +110,7 @@ export interface MutationResultAction {
   type: 'APOLLO_MUTATION_RESULT';
   result: ExecutionResult;
   document: DocumentNode;
-  operationName: string;
+  operationName: string | null;
   variables: Object;
   mutationId: string;
   extraReducers?: ApolloReducer[];
@@ -136,7 +136,7 @@ export interface UpdateQueryResultAction {
   type: 'APOLLO_UPDATE_QUERY_RESULT';
   variables: any;
   document: DocumentNode;
-  operationName: string;
+  operationName: string | null;
   newResult: Object;
 }
 
@@ -159,7 +159,7 @@ export interface SubscriptionResultAction {
   subscriptionId: number;
   variables: Object;
   document: DocumentNode;
-  operationName: string;
+  operationName: string | null;
   extraReducers?: ApolloReducer[];
 }
 
@@ -171,7 +171,7 @@ export interface DataWrite {
   rootId: string;
   result: any;
   document: DocumentNode;
-  operationName: string;
+  operationName: string | null;
   variables: Object;
 }
 

--- a/src/data/mutationResults.ts
+++ b/src/data/mutationResults.ts
@@ -5,7 +5,7 @@ import {
 // This is part of the public API, people write these functions in `updateQueries`.
 export type MutationQueryReducer = (previousResult: Object, options: {
   mutationResult: Object,
-  queryName: Object,
+  queryName: string | null,
   queryVariables: Object,
 }) => Object;
 

--- a/src/queries/getFromAST.ts
+++ b/src/queries/getFromAST.ts
@@ -59,8 +59,8 @@ string in a "gql" tag? http://docs.apollostack.com/apollo-client/core.html#gql`)
   });
 }
 
-export function getOperationName(doc: DocumentNode): string {
-  let res: string = '';
+export function getOperationName(doc: DocumentNode): string | null {
+  let res: string | null = null;
   doc.definitions.forEach((definition) => {
     if (definition.kind === 'OperationDefinition' && definition.name) {
       res = definition.name.value;

--- a/src/transport/networkInterface.ts
+++ b/src/transport/networkInterface.ts
@@ -33,7 +33,7 @@ export interface Request {
   debugName?: string;
   query?: DocumentNode;
   variables?: Object;
-  operationName?: string;
+  operationName?: string | null;
   [additionalKey: string]: any;
 }
 
@@ -43,7 +43,7 @@ export interface PrintedRequest {
   debugName?: string;
   query?: string;
   variables?: Object;
-  operationName?: string;
+  operationName?: string | null;
 }
 
 export interface NetworkInterface {

--- a/test/getFromAST.ts
+++ b/test/getFromAST.ts
@@ -211,6 +211,15 @@ describe('AST utility functions', () => {
     assert.equal(operationName, 'nameOfMutation');
   });
 
+  it('should return null if the query does not have an operation name', () => {
+    const query = gql`
+      {
+        fortuneCookie
+      }`;
+    const operationName = getOperationName(query);
+    assert.equal(operationName, null);
+  });
+
   it('should throw if type definitions found in document', () => {
     const queryWithTypeDefination = gql`
       fragment authorDetails on Author {

